File: other_algorithms/zigzag/zigzag.cpp
#include <iostream>
#include <vector>
#include <sstream>

using namespace std;


vector<int> get_value(){
    std::string input_str;
    vector<int> element;
    
    getline(std::cin, input_str);
    stringstream ss(input_str);
    std::string temp;
    char delimiter = ',';
    while (getline(ss,temp, delimiter))
    {
        element.push_back(std::stoi(temp));
    }

    return element;
}


int main(){
    std::cout << "Enter elemnts array_1(separate elements with commas): ";
    vector<int> array_1 = get_value();
    
    std::cout << "Enter elemnts array_2(separate elements with commas): ";
    vector<int> array_2 = get_value();
    vector<int> result;

    int len_1 = array_1.size();
    int len_2 = array_2.size();
    if(len_1 >= len_2){
        for (int i = 0; i < len_1; i++) {
            
            if(i == len_2){
                std::vector<int> slice_vector = std::vector<int>(array_1.begin() + i, array_1.end());

                // c++23
                result.append_range(slice_vector);
                break;
            }
    
            result.push_back(array_1[i]);
            result.push_back(array_2[i]);
        }
    }
    else
    {
        for (int i = 0; i < len_2; i++) {
            
            if(i == len_1){
                std::vector<int> slice_vector = std::vector<int>(array_2.begin() + i, array_2.end());

                // c++23
                result.append_range(slice_vector);
                break;
            }
    
            result.push_back(array_1[i]);
            result.push_back(array_2[i]);
        }
    }
    

    for (auto i : result) {
        cout << i << " ";
    }
    

    return 0;
}

File: other_algorithms/zigzag/zigzag.py
from typing import List


# simple method
def zigzag(array_1: List, array_2: List) -> List:
    array = []
    if len(array_2) == 0:
        return array_1

    len_1 = len(array_1)
    len_2 = len(array_2)
    
    if len_1 >= len_2:
        for i in range(len_1):
            if i == len_2:
                array.extend(array_1[i:])
                break
            array.append(array_1[i])
            array.append(array_2[i])
    else:
        for i in range(len_2):
            if i == len_1:
                array.extend(array_2[i:])
                break
            array.append(array_1[i])
            array.append(array_2[i])
    
    return array
            

# Best method
class ZigZag:
    def __init__(self, array_1: List, array_2: List):
        self.queue = [array_1, array_2]
    
    def next(self):
        value = self.queue.pop(0)
        item = value.pop(0)
        
        if value:
            self.queue.append(value)
        return item
    
    def has_next(self):
        if self.queue:
            return True
        return False

 
array1 = list(map(int, input("Enter elemnts array_1: ").split(",")))
array2 = list(map(int, input("Enter elemnts array_2: ").split(",")))
result = zigzag(array1, array2)
print(f"Result: {result}")

zz_obj = ZigZag(array1, array2)
while zz_obj.has_next():
    print(zz_obj.next(), end=',')
    